vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,314 @@
+ * A declaration of a variable and assignment of its value is simply done as a property in the variables YAML file.
+ * <p>
+ * The format follows <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution">Cloud Foundry standards</a>.
+ * <p>
+ * The step is activated by the presence of both a {@code manifest.yml} and a variables file. Names of both files are configurable.
+ * <p>
+ * Usage: yamlSubstituteVariables manifestFile: "manifest.yml", variablesFile: "manifest-variables.yml"
+ *
+ * @param arguments - the map of arguments.
+ */
+@GenerateDocumentation
+void call(Map<String, String> arguments) {
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: arguments) {
+        def script = checkScript(this, arguments)  ?: this
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+                                        .loadStepDefaults()
+                                        .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+                                        .mixinStageConfig(script.commonPipelineEnvironment, arguments.stageName ?: env.STAGE_NAME, STEP_CONFIG_KEYS)
+                                        .mixin(arguments, PARAMETER_KEYS)
+                                        .use()
+
+        String manifestFilePath = config.manifestFile ?: "manifest.yml"
+        String variablesFilePath = config.variablesFile ?: "manifest-variables.yml"
+        String outputFilePath = config.outputManifestFile ?: manifestFilePath
+
+        // Parameter for tests only. No public API.
+        // This is to skip manifest file deletion in tests, which
+        // would be a nuisance, as it alters the sources repository.
+        Boolean skipDeletion = arguments.skipDeletion ?: false
+
+        File manifestFile = new File(manifestFilePath)
+        File variablesFile = new File(variablesFilePath)
+
+        if (!manifestFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find YAML file at ${manifestFilePath}. Skipping variable substitution."
+            return
+        }
+
+        if (!variablesFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find variable substitution file at ${variablesFilePath}. Skipping variable substitution."
+            return
+        }
+
+        def manifestData = null;
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            manifestData = readYaml file: manifestFilePath
+            echo "[YamlSubstituteVariables] Loaded manifest at ${manifestFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest file at ${manifestFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
```

Comment:
1.) multiple errors one after the other does not work. An error results in an `hudons.AbortException`. The second `error` statement will never be invoked.
2.) The `debug` output provides the same information like the second error statement. Hence the `debug` statement not needed.
3.) What would be wrong with simply letting the original exception bubble up ...

Reply:
Whoops! :) Good catch. That really does not make sense.
If bubbling up the exception is properly failing the build, then I would prefer that option actually. I just was not sure how Piper would react to it.
I will change that.

Reply:
**Update** - It's done. I pushed a new commit that resolves this.